### PR TITLE
fixed a bunch of tests on darwin

### PR DESCRIPTION
**Pull request by ahl**
Everything except for unionfs is now passing. Still chasing down some bugs there, but wanted to see if there was interest before continuing.

Found a legit bug in unionfs:
When the Finder sent it an unexpected getxattr unionfs responded ENOSYS because the subordinate filesystem returned it. Returning ENOSYS causes FUSE to *never* send more getxattrs. We can intercept that and replace it with ENODATA.

### CODE REVIEW

File: fuse/server.go [hunk 1]
```diff
@@ -25,6 +25,9 @@ const (
 // Server contains the logic for reading from the FUSE device and
 // translating it to RawFileSystem interface calls.
 type Server struct {
+	versionMajor uint32
```
Comment (hanwen) [inline fuse/server.go]: this is already available in the kernelSettings member.

File: fuse/test/loopback_darwin_test.go [hunk 2]
```diff
@@ -10,9 +10,14 @@ import (
...
 	s.Fsid = empty.Fsid
-	// s.Spare = empty.Spare
-	// TODO - figure out what this is for.
-	s.Flags = 0
+	s.Owner = empty.Owner
+	s.Flags = empty.Flags
+	s.Fssubtype = empty.Fssubtype
+	s.Fstypename = empty.Fstypename
+	s.Mntonname = empty.Mntonname
+	s.Mntfromname = empty.Mntfromname
```
Comment (hanwen) [inline fuse/test/loopback_darwin_test.go]: maybe it's easier to init to empty and copy the fields that FUSE can set?

File: unionfs/autounion_test.go [hunk 3]
```diff
@@ -142,14 +144,22 @@ func TestAutoFsSymlink(t *testing.T) {
...
-	fi, _ = os.Lstat(wd + "/mnt/manual1")
-	if fi != nil {
-		t.Error("Should not have file:", fi)
-	}
-
-	_, err = ioutil.ReadDir(wd + "/mnt/config")
-	if err != nil {
-		t.Fatalf("ReadDir failed: %v", err)
+	// Note that FUSE API 7.11 introduced the notification support required to cause the node to be immediately removed. This check will fail in earlier versions.
+	if _, minor := server.Version(); minor >= 11 {
```
Comment (hanwen) [inline unionfs/autounion_test.go]: it's unlikely that major will change from 7 , but maybe easier to add 

  func (in *InitIn) SupportsOpcode(opcode int)

and/or  

  func (in *InitIn) SupportsFeature(flag int)

File: unionfs/unionfs.go [hunk 4]
```diff
@@ -712,7 +712,12 @@ func (fs *unionFS) GetXAttr(name string, attr string, context *fuse.Context) ([]
 	}
 	r := fs.getBranch(name)
 	if r.branch >= 0 {
-		return fs.fileSystems[r.branch].GetXAttr(name, attr, context)
+		value, code := fs.fileSystems[r.branch].GetXAttr(name, attr, context)
+		// We intercept ENOSYS and replace it with ENODATA; returning ENOSYS would cause FUSE to never to GetXAttr() again.
+		if code == fuse.ENOSYS {
+			return nil, fuse.ENODATA
+		}
+		return value, code
```
Comment (hanwen) [inline unionfs/unionfs.go]: this can be a separate CL. ENOSYS should be intercepted in opcode.go (there should be handling for ENOSYS in connection with the DisableXattr option.)

File: README [hunk 5]
```diff
@@ -104,8 +104,7 @@ go-fuse works somewhat on OSX. Known limitations:
 * OSX has trouble with concurrent reads from the FUSE device, leading
```
Comment (hanwen) [inline README]: thanks for looking into this. It's much appreciated!

File: unionfs/autounion_test.go (hunk 3 above)
Comment (ahl) [inline unionfs/autounion_test.go]: Good idea. I'd suggest that it belongs in Server rather than InitIn

File: fuse/test/loopback_darwin_test.go (hunk 2 above)
Comment (ahl) [inline fuse/test/loopback_darwin_test.go]: Good idea.

File: unionfs/unionfs.go (hunk 4 above)
Comment (ahl) [inline unionfs/unionfs.go]: I think this makes more sense in unionfs. Typically a filesystem will either implement GetXAttr or they'll return ENOSYS in which case the existing behavior is desirable. I don't think you want to force users to know they should set DisableXAttr to avoid a bunch of repeated useless requests.

Can you imagine a scenario where what you're describing would be beneficial? This issue would seem to be one with unionfs not one that applies more generally.

File: unionfs/autounion_test.go (hunk 3 above)
Comment (hanwen) [inline unionfs/autounion_test.go]: It's all keyed on the InitIn data, though, so I think it should rather be on InitIn.

File: unionfs/unionfs.go (hunk 4 above)
Comment (hanwen) [inline unionfs/unionfs.go]: Support xattr is a property of the entire mount point, since FUSE can't decide which parts of the filesystem do or do not support it. Hence the support should also be where the mountpoint is handled.

This is already partly in place, see 7b281480b21a59f54a0c86451e7de13cd9be736f.   - it just needs to transform ENOSYS into ENODATA there.

File: unionfs/autounion_test.go (hunk 3 above)
Comment (ahl) [inline unionfs/autounion_test.go]: so... turn `kernelSettings` into `KernelSettings`?

File: unionfs/unionfs.go (hunk 4 above)
Comment (ahl) [inline unionfs/unionfs.go]: It seems like either ENOSYS should be converted into ENODATA in opcode.go or that the approach in https://github.com/hanwen/go-fuse/commit/7b281480b21a59f54a0c86451e7de13cd9be736f should be continued (i.e. change ENOSYS to ENODATA in fuse/pathfs/default.go).

If we choose the former, then I'd suggest reverting the previous changes in fuse/defaultraw.go and fuse/nodefs/defaultnode.go.

Let me know.

**Comment by ahl** [comment]
pushed an updated diff for review.

File: fuse/server.go [hunk 6]
```diff
@@ -64,8 +61,12 @@ func (ms *Server) SetDebug(dbg bool) {
 	ms.opts.Debug = dbg
 }
 
-func (ms *Server) Version() (major, minor uint32) {
-	return ms.versionMinor, ms.versionMinor
+func (ms *Server) SupportsNotify(op int) bool {
+	switch op {
+	case NOTIFY_INVAL_ENTRY:
+		return ms.kernelSettings.Major >= 7 && ms.kernelSettings.Minor >= 12
```
Comment (ahl) [inline fuse/server.go]: confirming the version when support was added:
https://github.com/osxfuse/kext/blob/5639d3e9c9aa32c532a0faef562d6c459a485f8d/osxfuse/fuse_ipc.h#L511

File: unionfs/unionfs.go (hunk 4 above)
Comment (hanwen) [inline unionfs/unionfs.go]: I think opcode.go is the best place. It guarantees that this stops biting us.

File: fuse/server.go (hunk 6 above)
Comment (hanwen) [inline fuse/server.go]: Let me provide something for query what the server supports.

(to answer another question: there is already a KernelSEttings() method.)

Comment (hanwen) [inline fuse/server.go]: see https://github.com/hanwen/go-fuse/commit/3cf08068e8f6ff0c7961023efdf318b44a3ba719

Comment (ahl) [inline fuse/server.go]: Seems needlessly indirect, but, fine, I guess I'll use it now.

File: unionfs/unionfs.go (hunk 4 above)
Comment (ahl) [inline unionfs/unionfs.go]: I'd note that that seems like the wrong methodology and would potentially create surprises for users, but fine. Shall I revert the incomplete work in https://github.com/hanwen/go-fuse/commit/7b281480b21a59f54a0c86451e7de13cd9be736f and change those ENODATA returns to ENOSYS?

**Comment by ahl** [comment]
updated to use your new interface.

**Comment by ahl** [comment]
all tests are now passing on darwin; finally!

File: fuse/server.go [hunk 7]
```diff
@@ -560,13 +552,13 @@ func (ms *Server) EntryNotify(parent uint64, name string) Status {
 
 // SupportsVersion returns true if the kernel supports the given
 // protocol version or newer.
-func (in *InitIn) SupportsVersion(maj, min uint32) bool {
+func (in InitIn) SupportsVersion(maj, min uint32) bool {
```
Comment (hanwen) [inline fuse/server.go]: why? this is making a copy of InitIn now.

**Comment by hanwen** [comment]
I've squashed this and cherry-picked (except the InitIn change). Thanks!

File: fuse/server.go (hunk 7 above)
Comment (ahl) [inline fuse/server.go]: Because `KernelSettings()` returns a non-pointer. The zipfs and unionfs tests don't build right now on Darwin; do they build and work on Linux?

Comment (hanwen) [inline fuse/server.go]: ah, you're right. I've changed KernelSettings to return a pointer.

Comment (ahl) [inline fuse/server.go]: I didn't take that approach because I figured the goal was to give the user something that wouldn't mutate the original.
